contrib/epee/src/net_ssl.cpp

```diff
@@ -0,0 +1,280 @@
+// 
+// All rights reserved.
+// 
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+// 
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+// 
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+// 
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <boost/asio/ssl.hpp>
+#include <openssl/ssl.h>
+#include <openssl/pem.h>
+#include "misc_log_ex.h"
+#include "net/net_ssl.h"
+
+#undef MONERO_DEFAULT_LOG_CATEGORY
+#define MONERO_DEFAULT_LOG_CATEGORY "net.ssl"
+
+// openssl genrsa -out /tmp/KEY 4096
+// openssl req -new -key /tmp/KEY -out /tmp/REQ
+// openssl x509 -req -days 999999 -sha256 -in /tmp/REQ -signkey /tmp/KEY -out /tmp/CERT
+
+namespace epee
+{
+namespace net_utils
+{
+
+// https://stackoverflow.com/questions/256405/programmatically-create-x509-certificate-using-openssl
+bool create_ssl_certificate(EVP_PKEY *&pkey, X509 *&cert)
+{
+  MGINFO("Generating SSL certificate");
+  pkey = EVP_PKEY_new();
+  if (!pkey)
+  {
+    MERROR("Failed to create new private key");
+    return false;
+  }
+  RSA *rsa = RSA_generate_key(4096, RSA_F4, NULL, NULL);
+  if (!rsa)
+  {
+    MERROR("Error generating RSA private key");
+    EVP_PKEY_free(pkey);
+    return false;
+  }
+  EVP_PKEY_assign_RSA(pkey, rsa);
+
+  cert = X509_new();
+  if (!cert)
+  {
+    MERROR("Failed to create new X509 certificate");
+    EVP_PKEY_free(pkey);
+    return false;
+  }
+  ASN1_INTEGER_set(X509_get_serialNumber(cert), 1);
+  X509_gmtime_adj(X509_get_notBefore(cert), 0);
+  X509_gmtime_adj(X509_get_notAfter(cert), 3600 * 24 * 365 * 10); // 10 years
+  X509_set_pubkey(cert, pkey);
+  X509_NAME *name = X509_get_subject_name(cert);
```

Comment:
This cert was just created, so its subject_name is empty. Do we want to put some kind of non-empty name in here?

Reply:
That's a good question. Perhaps something with boost::asio::hostname() ?

Reply:
I would definitely not want my hostname to leak through a SSL certificate. What is that name used for in practice ?

Reply:
The name in a server cert usually corresponds to its hostname, and clients will verify that the name in the cert matches the name the client tried to connect to. This is one of the typical means of detecting spoofing attacks. Before progressing any further, we should define what is the threat that SSL is being used to protect against?

Reply:
AFAIK that name matching's only useful when using CAs. Here, we don't, we give monero the cert we expect to see. No spoofing possible (unless the cert's private key was compromised).

SSL is now used only for RPC. The point is to hide the data from network level snoopers (or people who'd modify the data on the way). And generally add some more encrypted traffic to piss off grab-everything snoops :)

For those people who use a round robin RPC server, they'll have to be set to accept any cert, but they're not expecting to connect to any particular server anyway, so nothing lost. They still gain privacy from their ISP (unless that ISP actively MITMs).

Reply:
> we give monero the cert we expect to see.

I'm missing something here, then. This code generates a cert on the fly, on the monerod. So how does the client "expect" to see this? There's no protection against MITM at all, since any 3rd party could intercept and use their own generated cert. If active MITM is not something we're defending against, we can just skip all of this and use an anonymous Diffie-Hellman ciphersuite for the SSL session.

Reply:
It seems to be only used when `certificate_path` is empty, so when no certificate was generated, this code protects against passive eavesdropping, when given, active MITM.

Reply:
Yes. If you use --*-allow-any-cert, it will accept any. Otherwise it will require a cert that's in the --*-allowed-certificates list.